ui/appui-layout-react/src/appui-layout-react/state/NineZoneStateReducer.ts

```diff
@@ -366,37 +371,60 @@ export function NineZoneStateReducer(
       const popoutWidget = state.popoutWidgets.byId[action.id];
       const widget = getWidgetState(state, action.id);
       const home = popoutWidget.home;
-      const panel = state.panels[home.side];
-      let widgetPanelSectionId = home.widgetId;
-      let homeWidgetPanelSection;
-      if (widgetPanelSectionId) {
-        homeWidgetPanelSection = state.widgets[widgetPanelSectionId];
-      } else {
-        widgetPanelSectionId = getWidgetPanelSectionId(
-          panel.side,
+
+      state = removeWidget(state, widget.id);
+
+      // Determine existing widget.
+      let existingWidget: WidgetState | undefined;
+      if (home.widgetId in state.widgets) {
+        existingWidget = state.widgets[home.widgetId];
+      } else if (isPanelWidgetRestoreState(home)) {
+        const panel = state.panels[home.side];
+        const panelSectionId = getWidgetPanelSectionId(
+          home.side,
           home.widgetIndex
         );
-        homeWidgetPanelSection = state.widgets[widgetPanelSectionId];
+        existingWidget = state.widgets[panelSectionId];
+        if (!existingWidget && panel.widgets.length >= panel.maxWidgetCount) {
+          const sectionIndex = Math.min(
+            panel.maxWidgetCount - 1,
+            home.widgetIndex
+          );
+          const sectionId = panel.widgets[sectionIndex];
+          existingWidget = state.widgets[sectionId];
+        }
       }
 
-      state = removeWidget(state, popoutWidget.id);
-      if (homeWidgetPanelSection) {
-        const tabs = [...homeWidgetPanelSection.tabs, ...widget.tabs];
-        state = updateWidgetState(state, homeWidgetPanelSection.id, {
+      // Add to an existing widget.
+      if (existingWidget) {
+        const tabs = [...existingWidget.tabs, ...widget.tabs];
+        return updateWidgetState(state, existingWidget.id, {
           tabs,
         });
-      } else {
-        // if widget panel section was removed because it was empty insert it
-        const sectionIndex = widgetPanelSectionId.endsWith("End") ? 1 : 0;
-        state = insertPanelWidget(
+      }
+
+      // Insert a new panel section.
+      if (isPanelWidgetRestoreState(home)) {
```

Comment:
There is a bug here where we try to use the `widgetId` that we already validated did not exist instead of the correct `getWidgetPanelSectionId` that was saved and should be created, and this cause the poped out widget to always go back to `start` section if the `end` section do not exist. (It works well if there is already a widget in the end section though.)

Reply:
Fixed by removing the substring check that was copied from appui-react.